### ISSUE TITLE
Issue with the transformers optimizer when arg --input_int32 is used

### ISSUE THREAD

**Issue by jplu** [microsoft/onnxruntime#8853]
**Describe the bug**
The error `Unknown model file format version.` everytime a model is optimized with the argument `--input_int32` is passed to the `onnxruntime.transformers.optimizer` script.

**Urgency**
None

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux 21.04
- ONNX Runtime installed from (source or binary): binary from pypi
- ONNX Runtime version: 1.8.1
- Python version: 3.8.10
- CUDA/cuDNN version: 11.2
- GPU model and memory: RTX 2080 Super mobile / Max-Q 8Go RAM
- transformers version: 4.9.2

**To Reproduce**
Steps to reproduce directly from the terminal:
```
python -m transformers.onnx --model=bert-base-cased onnx/bert-base-cased/
python -m onnxruntime.transformers.optimizer --input onnx/bert-base-cased/model.onnx --output onnx/bert-base-cased/model_opt.onnx --disable_embed_layer_norm --model_type bert --num_heads 12 --input_int32 --hidden_size 768 --float16 --opt_level 99
python -m onnxruntime.transformers.compare_bert_results --baseline_model onnx/bert-base-cased/model.onnx --optimized_model onnx/bert-base-cased/model_opt.onnx --input_ids input_ids --segment_ids token_type_ids --input_mask attention_mask --batch_size 1 --sequence_length 128 --samples 100
```

**Expected behavior**
Having the possibility to properly load an optimized model when it has been optimized with the `--input_int32` argument.

**Additional context**
Here the full error stack:
```
Traceback (most recent call last):
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/site-packages/onnxruntime/transformers/compare_bert_results.py", line 175, in <module>
    main()
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/site-packages/onnxruntime/transformers/compare_bert_results.py", line 169, in main
    run_test(args.baseline_model, args.optimized_model, args.output_dir, args.batch_size, args.sequence_length,
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/site-packages/onnxruntime/transformers/compare_bert_results.py", line 108, in run_test
    treatment_results, treatment_latency, treatment_output_names = run_model(optimized_model,
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/site-packages/onnxruntime/transformers/compare_bert_results.py", line 36, in run_model
    session = create_session(model_path, use_gpu, intra_op_num_threads, graph_optimization_level)
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/site-packages/onnxruntime/transformers/bert_perf_test.py", line 84, in create_session
    session = onnxruntime.InferenceSession(model_path, sess_options, providers=execution_providers)
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 283, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 310, in _create_inference_session
    sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model)
onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Load model from onnx/bert-base-cased/model_opt.onnx failed:/onnxruntime_src/onnxruntime/core/graph/model.cc:111 onnxruntime::Model::Model(onnx::ModelProto&&, const PathString&, const IOnnxRuntimeOpSchemaRegistryList*, const onnxruntime::logging::Logger&) Unknown model file format version.
```

Thanks in advance for any hints.

**Comment by tianleiwu** [comment]
@jplu, compare_bert_results tool assumes that two ONNX models has same input data type. If --input_int32 is used, input_ids in original onnx model is int64, and optimized model has int32 inputs.

As walkaround, you can generate a model without --input_int32 for parity verification. Once the verification is done, you can add --input_int32 to generate the final model for performance test etc.

Here is example commands to run from script of master branch:
```
git clone https://github.com/microsoft/onnxruntime
cd onnxruntime/onnxruntime/python/tools/transformers
python -m transformers.onnx --model=bert-base-cased onnx/bert-base-cased/
python optimizer.py --input onnx/bert-base-cased/model.onnx --output onnx/bert-base-cased/model_opt.onnx --model_type bert --float16
python compare_bert_results.py --baseline_model onnx/bert-base-cased/model.onnx --optimized_model onnx/bert-base-cased/model_opt.onnx --input_ids input_ids --segment_ids token_type_ids --input_mask attention_mask --batch_size 1 --sequence_length 128 --samples 100
python optimizer.py --input onnx/bert-base-cased/model.onnx --output onnx/bert-base-cased/model_int32.onnx --model_type bert --float16 --input_int32
python bert_perf_test.py --model onnx/bert-base-cased/model_int32.onnx --use_gpu -b 1 -s 128 -t 100 --opt_level 99
```

Example outputs:
```
>python optimizer.py --input onnx/bert-base-cased/model.onnx --output onnx/bert-base-cased/model_opt.onnx --model_type bert --float16
               apply: Fused LayerNormalization count: 25
               apply: Fused Gelu count: 12
               apply: Fused SkipLayerNormalization count: 24
               apply: Fused Attention count: 12
         prune_graph: Graph pruned: 0 inputs, 0 outputs and 5 nodes are removed
               apply: Fused EmbedLayerNormalization(with mask) count: 1
         prune_graph: Graph pruned: 0 inputs, 0 outputs and 4 nodes are removed
         prune_graph: Graph pruned: 0 inputs, 0 outputs and 0 nodes are removed
               apply: Fused BiasGelu count: 12
               apply: Fused SkipLayerNormalization(add bias) count: 24
            optimize: opset verion: 12
  save_model_to_file: Sort graphs in topological order
  save_model_to_file: Output model to onnx/bert-base-cased/model_opt.onnx
get_fused_operator_statistics: Optimized operators:{'EmbedLayerNormalization': 1, 'Attention': 12, 'Gelu': 0, 'FastGelu': 0, 'BiasGelu': 12, 'LayerNormalization': 0, 'SkipLayerNormalization': 24}
                main: The model has been fully optimized.

>python compare_bert_results.py --baseline_model onnx/bert-base-cased/model.onnx --optimized_model onnx/bert-base-cased/model_opt.onnx --input_ids input_ids --segment_ids token_type_ids --input_mask attention_mask --batch_size 1 --sequence_length 128 --samples 100
WARNING: 100 out of 100 results NOT passed for thresholds (rtol=0.001, atol=0.0001).
maximum absolute difference=0.008105754852294922
maximum relative difference=1947.8863525390625
```
The threshold is for fp32 model. For fp16 model, I think the maximum absolute difference is expected (since the resolution of fp16 is around 0.001).

**Comment by jplu** [comment]
Thanks @tianleiwu I tried what you suggested and I get the same error with the command:
```
python -m onnxruntime.transformers.bert_perf_test --model  onnx/bert-base-cased/model_opt.onnx --opt_level 99 -b 1 -s 128 -t 100 --input_ids_name input_ids --input_mask_name attention_mask --segment_ids token_type_ids --use_gpu
```

```
test setting TestSetting(batch_size=1, sequence_length=128, test_cases=10, test_times=100, use_gpu=True, intra_op_num_threads=None, seed=3, verbose=False)
Generating 10 samples for batch_size=1 sequence_length=128
Process Process-2:
Traceback (most recent call last):
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/site-packages/onnxruntime/transformers/bert_perf_test.py", line 116, in run_one_test
    session = create_session(model_setting.model_path, test_setting.use_gpu, intra_op_num_threads,
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/site-packages/onnxruntime/transformers/bert_perf_test.py", line 84, in create_session
    session = onnxruntime.InferenceSession(model_path, sess_options, providers=execution_providers)
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 283, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 310, in _create_inference_session
    sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model)
onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Load model from onnx/encoder/model_opt.onnx failed:/onnxruntime_src/onnxruntime/core/graph/model.cc:111 onnxruntime::Model::Model(onnx::ModelProto&&, const PathString&, const IOnnxRuntimeOpSchemaRegistryList*, const onnxruntime::logging::Logger&) Unknown model file format version.
```

The exact same error happen also when I try to just open the model:
```
import onnxruntime as ort

ort_session = ort.InferenceSession("onnx/bert-base-cased/model_opt.onnx")
```

What you suggest apparently doesn't seems to work :(  any other walkaround, or is that a real bug?

**Comment by tianleiwu** [comment]
@jplu, 

`Unknown model file format version` is probably caused by related packages (like protobuf). You can check related packages like the following Python script:
```
import pkg_resources
installed_packages = pkg_resources.working_set
related_packages = ['flatbuffers', 'numpy', 'onnx', 'onnxconverter-common', 'onnxruntime-gpu', 'onnxruntime', 'ort-nightly-gpu', 'ort-nightly', 'protobuf', 'sympy', 'torch', 'transformers']
related_packages_list = sorted(["%s==%s" % (i.key, i.version) for i in installed_packages if i.key in related_packages])
print(related_packages_list)
```
In my machine, the output is like
```
['flatbuffers==2.0', 'numpy==1.19.5', 'onnx==1.9.0', 'onnxconverter-common==1.9.0', 'onnxruntime-gpu==1.8.1', 'protobuf==3.17.3', 'sympy==1.5', 'torch==1.9.0+cu111', 'transformers==4.9.2']
```

BTW, the following command works in my machine (you will need run optimizer.py in master branch to get fully optimized model).
```
>python -m onnxruntime.transformers.bert_perf_test --model transformers/onnx/bert-base-cased/model_opt.onnx --opt_level 99 -b 1 -s 128 -t 100 --input_ids_name input_ids --input_mask_name attention_mask --segment_ids token_type_ids --use_gpu

test setting TestSetting(batch_size=1, sequence_length=128, test_cases=10, test_times=100, use_gpu=True, intra_op_num_threads=None, seed=3, verbose=False)
Generating 10 samples for batch_size=1 sequence_length=128
Running test: model=model_opt.onnx,graph_optimization_level=ENABLE_ALL,intra_op_num_threads=12,batch_size=1,sequence_length=128,test_cases=10,test_times=100,use_gpu=True
Average latency = 8.89 ms, Throughput = 112.44 QPS
```

**Comment by jplu** [comment]
Here is my output:
```
['flatbuffers==2.0', 'numpy==1.20.3', 'onnx==1.10.1', 'onnxconverter-common==1.8.1', 'onnxruntime-gpu==1.8.1', 'protobuf==3.17.3', 'sympy==1.8', 'torch==1.9.0+cu111', 'transformers==4.9.2']
```

> BTW, you will need run optimizer.py in master branch (instead of 1.8.1) to get fully optimized model, since 1.8.1 does not
support the latest transformers.

I can use only pypi packages for now because of some reproducible reasons.

> Also, remove --opt_level 99 in using optimizer.py, otherwise the model cannot be fully optimized.

Thanks for the hint! Good to know! Might be worth to specify this in the doc. And, by curiosity, why with this option the model cannot be fully optimized?

**Comment by tianleiwu** [comment]
@jplu, I reproduced the problem with onnx 1.10.1. You can downgrade onnx to 1.9.0 to walkaround the issue.

--opt_level 99 will use onnxruntime to do graph optimization (written in C++). It has some extra optimizations that could change the graph, which will cause no match in attention fusion in python later.

**Comment by jplu** [comment]
>@jplu, I reproduced the problem with onnx 1.10.1. You can downgrade onnx to 1.9.0 to walkaround the issue. When 1.8.1 was released, onnx was at 1.9.0. I will investigate the root cause later.

Perfect, thanks a lot @tianleiwu for the walkaround. Happy to have helped to find a potential bug :)

>--opt_level 99 will use onnxruntime to do graph optimization (written in C++). It has some extra optimizations that could change the graph, which will cause no match in attention fusion in python later.

Humm I see. Do you suggest to use only `--opt_level` set to 1 or 2? Or no `--opt_level` at all?

**Comment by jplu** [comment]
It seems there might be another issue, even though the model can now be properly loaded it raises the following error when the inputs are in `int32`:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jplu/anaconda3/envs/indexing/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 188, in run
    return self._sess.run(output_names, input_feed, run_options)
onnxruntime.capi.onnxruntime_pybind11_state.InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Unexpected input data type. Actual: (tensor(int32)) , expected: (tensor(int64))
```
The command used to get the model is still similar to what I proposed:
```
python -m onnxruntime.transformers.optimizer --input onnx/bert-base-cased/model.onnx --output onnx/bert-base-cased/model_opt.onnx --disable_embed_layer_norm --model_type bert --num_heads 12 --hidden_size 768 --float16 --input_int32
```

Here also the small piece of code I used:
```
import onnxruntime as ort
from transformers import AutoTokenizer

ort_session = ort.InferenceSession("onnx/bert-base-cased/model_opt.onnx")
tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")
inputs = tokenizer("Hello there!!", return_tensors="np")
inputs = {key: val.astype(np.int32) for key, val in inputs.items()}
ort_session.run(["last_hidden_state", "pooler_output"], dict(inputs))
```

**Comment by tianleiwu** [comment]
@jplu, the --input_int32 is designed for model with EmbedLayerNormalization. If you disable it (--disable_embed_layer_norm), only attention mask will be converted to int32, and other two inputs are still int64. You can open the onnx file with Netron and inspect the input data types.

**Comment by jplu** [comment]
Thanks a lot for the explanation!

**Comment by jplu** [comment]
Apparently Netron shows that `input_ids` and `token_type_ids` are in int32 but `attention_mask` is still in int64. Here some screenshots
![Capture d’écran de 2021-08-31 10-59-45](https://user-images.githubusercontent.com/959590/131474303-1e145c00-ca22-4806-9e7d-d84b52338c6f.png)
![Capture d’écran de 2021-08-31 11-00-22](https://user-images.githubusercontent.com/959590/131474319-d7f913f6-50f5-409e-957e-ae8257ea4b41.png)
